### PR TITLE
`exporter/loki`: handle multi-tenant use-cases

### LINKED ISSUE #3121: Loki exporter: handle multi-tenant use cases
**Is your feature request related to a problem? Please describe.**

The [lokiexporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/lokiexporter) has a `tenant_id` parameter that can set the `X-Scope-OrgID` HTTP header, but this sets a single static tenant.
We will have logs from 1000s of orgs coming in, and we were assuming that client devices would be setting `X-Scope-OrgID` HTTP header, and that the exporter would be passing this through to Loki, resulting in 1000s of separate per-org log streams in S3.

Currently this would require an exporter instance per tenant, which doesn't work.

**Describe the solution you'd like**

Enable passing OrgID either as the `X-Scope-OrgID` HTTP header or as a tenant "Attribute" on the spans/logs.

> Then within the Loki exporter we can use that attribute for setting the tenant header. The exporter groups logs into Loki LogStreams based on the cardinality of the labels, so the grouping would also need to account for the tenant attribute. It may make sense to combine this solution with the "[Group by Attributes](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/groupbyattrsprocessor)" processor in the pipeline. [slack](https://cloud-native.slack.com/archives/C01N6P7KR6W/p1617374532174000?thread_ts=1617261598.157500&cid=C01N6P7KR6W)

We want to send logs in the form of spans. I've created a separate issue for that: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3166

**Describe alternatives you've considered**

We don't at this time have an alternative for routing logs to S3 for long-term storage via the Collector.

**Additional context**

Related issues: 
- Access headers in processor (https://github.com/open-telemetry/opentelemetry-collector/issues/2401)
- Multi tenancy support proposal (https://github.com/open-telemetry/opentelemetry-collector/pull/2495)

### PR DESCRIPTION
**Pull request by jpkrohling**
This PR deprecates the existing tenant_id field in favor of a new tenant object, with a "source" and "value". The tenant information now may come from a resource attribute, from the context, or as a static value. Current setups using tenant_id are converted internally to "static".

This has been tested with a configuration similar to this:

```yaml
receivers:
  journald:
    directory: /var/log/journal/a04e3a44cdd740f88d6a7ae3bb8c70cf

exporters:
  logging:
    loglevel: debug
  loki:
    endpoint: http://localhost:3100/loki/api/v1/push
    tls:
      insecure: true
    labels:
      attributes:
        host.name: host_name 
    format: json
    tenant:
      source: attributes
      value: tenant

processors:
  attributes:
    actions:
    - action: insert
      key: host.name
      value: guarana
  resource:
    attributes:
    - action: insert
      key: tenant
      value: acme

extensions:

service:
  extensions:
  pipelines:
    logs:
      receivers: [journald]
      processors: [attributes, resource]
      exporters: [logging, loki]
```

Configuring a different tenant and a different data source on Grafana confirmed that only logs with the new tenant attribute show up there, now showing up on the first data source.

Fixes #3121

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
@dashpole, are you available to review this one?

File: exporter/lokiexporter/config.go [hunk 1]
```diff
@@ -32,19 +32,41 @@ type Config struct {
...
+	// should be the context key that holds the tenant information.
+	Value string `mapstruct:"value"`
 }
 
 func (c *Config) validate() error {
 	if _, err := url.Parse(c.Endpoint); c.Endpoint == "" || err != nil {
 		return fmt.Errorf("\"endpoint\" must be a valid URL")
 	}
 
+	if c.Tenant != nil {
```
Comment (TylerHelmuth) [inline exporter/lokiexporter/config.go:64]: What happens if both `Tenant` and `TenantID` are used at the same time?

File: unreleased/lokiexporter-multi-tenant.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+change_type: enhancement
+component: exporter/loki
```
Comment (TylerHelmuth) [inline unreleased/lokiexporter-multi-tenant.yaml:2]:
```suggestion
component: lokiexporter
```

Comment (jpkrohling) [reply unreleased/lokiexporter-multi-tenant.yaml:2]: There actually was a discussion in the past on which format folks preferred, and having the component type followed by the component's name (without the type) was chosen:

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6688

File: exporter/lokiexporter/config.go (hunk 1 above)
Comment (TylerHelmuth) [reply exporter/lokiexporter/config.go:64]: I like failing.  Maybe a little harsh, but mixing deprecated fields and their replacement fields feels like something we should help our users avoid.

Comment (jpkrohling) [reply exporter/lokiexporter/config.go:64]: PR updated

File: unreleased/lokiexporter-multi-tenant.yaml (hunk 2 above)
Comment (TylerHelmuth) [reply unreleased/lokiexporter-multi-tenant.yaml:2]: oh, we should start enforcing that then, the changelog is almost exclusively the other method haha

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**